### ISSUE TITLE
How about cobining gtm_proxy into postgres processes, as a child of postmaster? 

### ISSUE THREAD

**Issue by Jasonysli** [postgres-x2/postgres-x2#228]
gtm_proxy now is  a single componet, user has to construct high avaliable for it. And gtm_proxy only provide socket interface, it increases the network overhead of each transaction.If gtm_proxy was a child of postmater, then postmaster would take care of it when it crashed, and it can provide share memory interface for transactions, with very low overhead.

**Comment by koichi-szk** [comment]
Yes, this idea was discussed two years ago at PGCon.    Because current
communication between backends and gtm_proxy is based upon unix-domain
local socket, I hope efficiency itself will be similar to the case of
postmaster child.

Good thing is we can use many more PG infrastructure, such as WAL, recovery
and shared memories.   On the other hand, current gtm_proxy is based upon
pthread, where memory management, especially dynamic shared memory
management is much simpler than process shmem.

Also, I'm thinking that making GTM as postmaster child will be good idea to
use WAL, recovery log shipping for GTM recovery and GTM slave.

Another improvement will be to move current sequence management to
postmaster background worker, where we can resolve current XC sequence
restrictions.

To do this, porting to PG9.4 will be better which provide more flexible
background worker configuration.

If you're interested in this, we can create new sub project for this for
discussion and test code.

Let me know our idea.

Thanks;

---

Koichi Suzuki

2015-06-09 21:42 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
In our test, we found proxy will reduce the TPS by more than 40%, that is, with proxy, per coord TPS is 20K; without it , can be 35K. We tuned the performance issue, found it was caused by network latency. So we want to add proxy as a child of postmaster, and change the interface to shared memory queue, it will be much faster than unix domain socket.

**Comment by Jasonysli** [comment]
By doing this, another improvement is we can reduce the connetion number of gtm, in 20 node cluster, the connection to gtm may reduce to 40 or 50,other than 2000, then gtm will be more  efficiency.

**Comment by koichi-szk** [comment]
Gtm-prosy was introduced to reduce network workload, mainly by reducing
interaction to get a snapshot, as well as interaction between coordinator
and gtm.   In DBT-1 workload, we measured that total amount of data is
reduced to about a half.

By extra hop, yes we need more latency but gtm-proxy reduces total resource
for global transaction management.

Did you configure gtm-proxy at different node?   I always recommend to
configure gtm-proxy at the same node of coordinator/datanode.

Could you give me more detailed background how you get more throughput by
implementing gtm-proxy as postmaster child?

Regards;

---

Koichi Suzuki

2015-06-10 12:38 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
1、Yes, in our test, found proxy indeed can reduce the net workload. 
2、gtm proxy was configged in the same node as coord and datanode.
3、We will use shared memory lock free queue for gtm interfaces to reduce the interface latency.

**Comment by koichi-szk** [comment]
I have further question on 3.   Even though we use shared memory, we need some lock or synchronization mechanism to tell clients such as "the data is ready" or "you can issue further commands".

I think unix domain socket is implemented in such a manner.   Because this requires considerable change or almost rewrite current gtm/gtm_proxy code, we need to do some preliminary code work to show potential performance improvement.

Best Regards;

**Comment by Jasonysli** [comment]
Unix domain socket need copy data into kernel and then copy data out to receiver from kernel. So a whole msg sending needs two system calls, two memory copy action. If we use share memory queue, then we reduce system call action and memory copy action, so we can achieve better performance and lower latency. We will try to implement it next week and test the performance next month.

**Comment by koichi-szk** [comment]
Okay.   Before implementing full X2 infrastructure, I think it's better to write much simple test case to compare socket vs. shmem implementation and see the improvement.  If we share the idea works, then we can simplify the development process.

Regards;
